### ISSUE TITLE
state sync design for placement service

### ISSUE THREAD

**Issue by andypeng2015** [dapr/dapr#2897]
## In what area(s)?

 /area placement

## Ask your question here

dapr uses 3 phase lock to sync actor state, IMO it introduces single point of failure, it's also costly especially in cloud env, node is ephemeral, frequent lock unlock will impact performance, maybe batch sync will reduce the impact.

If there is batch process for state change sync then the cluster is not strong consistent but eventual consistency(node crashed but other nodes are not aware till batch sync to update state).

for akka cluster, it uses gossip protocol and failure detector for node joining and leaving, it converges fast.

what's the tradeoff behind this design? e.g. what's the big consideration for dapr not use gossip protocol for  eventual consistency

**Comment by artursouza** [comment]
@andypeng2015  Thanks for creating this issue.

Dapr uses Raft as the consensus algorithm to keep the state of membership for Actors. Every time a daprd instance (a.k.a. sidecar) goes online/offline, this state is updated among placement instances. Not all placement instances need to be up, only enough for quorum, so it is not a single point of failure.

Now, daprd instances do not invoke placement to get the membership state table, instead they keep a local copy. That copy is disseminated via a 3 phase lock, like you mentioned. That copy allows each daprd sidecar to know which instance hosts each actor id without querying placement.

Regarding performance, it would cause impact on performance if there is a large number of nodes going up or down in a small time window. We did discuss doing the updates in batch and this is an improvement we can make. Feel free to open a draft PR for that you have the time, please.

Finally, the membership state is eventual consistent since it is served from a local copy of the table and not from a quorum-based response.

I hope this helps clarify questions on the current implementation for placement service and membership table for actors.

@amanbha @youngbupark Please, add additional information here if I missed something.

**Comment by artursouza** [comment]
The current design was a safer choice to remove single point of failure from placement service by supporting multiple replicas.

It might be possible to implement using Gossip protocol. The initial challenge is how a sidecar can discover peer sidecars in standalone mode and Kubernetes mode. Placement service solves that problem since each sidecar "registers" with placement service on start up. 

/cc @yaron2

**Comment by andypeng2015** [comment]
@artursouza well understood, thanks.

I have 2 more clarifications before closing this ticket:
- how to achieve singleton actor (at most one actor instance for specific actor type) in the cluster? I do not see it in documentations
- Assume I set up one cluster behind a load balancer which assign client requests randomly, the receiving dapr instance received the request will route it to the correct dapr instance with the requested actor type, is my understanding correct? In this case, will the response be hopped back to the receiving dapr instance or directly to the client?

**Comment by yaron2** [comment]
> @artursouza well understood, thanks.
> 
> I have 2 more clarifications before closing this ticket:
> 
> * how to achieve singleton actor (at most one actor instance for specific actor type) in the cluster? I do not see it in documentations
> * Assume I set up one cluster behind a load balancer which assign client requests randomly, the receiving dapr instance received the request will route it to the correct dapr instance with the requested actor type, is my understanding correct? In this case, will the response be hopped back to the receiving dapr instance or directly to the client?

> how to achieve singleton actor (at most one actor instance for specific actor type) in the cluster? I do not see it in documentations

Singleton actors are the default and non-singleton actors are not allowed. At any given moment you can only have one actor instance of a specific type.

> Assume I set up one cluster behind a load balancer which assign client requests randomly, the receiving dapr instance received the request will route it to the correct dapr instance with the requested actor type, is my understanding correct? In this case, will the response be hopped back to the receiving dapr instance or directly to the client?

Your understanding is correct. your client will receive the response regardless of whether or not it was routed through a different Dapr host.

**Comment by yaron2** [comment]
> > @artursouza well understood, thanks.
> > I have 2 more clarifications before closing this ticket:
> > 
> > * how to achieve singleton actor (at most one actor instance for specific actor type) in the cluster? I do not see it in documentations
> > * Assume I set up one cluster behind a load balancer which assign client requests randomly, the receiving dapr instance received the request will route it to the correct dapr instance with the requested actor type, is my understanding correct? In this case, will the response be hopped back to the receiving dapr instance or directly to the client?
> 
> > how to achieve singleton actor (at most one actor instance for specific actor type) in the cluster? I do not see it in documentations
> 
> Singleton actors are the default and non-singleton actors are not allowed. At any given moment you can only have one actor instance of a specific type.
> 
> > Assume I set up one cluster behind a load balancer which assign client requests randomly, the receiving dapr instance received the request will route it to the correct dapr instance with the requested actor type, is my understanding correct? In this case, will the response be hopped back to the receiving dapr instance or directly to the client?
> 
> Your understanding is correct. your client will receive the response regardless of whether or not it was routed through a different Dapr host.

To clarify, an actor type can have multiple instances, and each instance has a unique I'd.

**Comment by andypeng2015** [comment]
@yaron2 thanks, so in this case how to achieve singleton actor (at most one actor instance for specific actor type) in the cluster?

**Comment by artursouza** [comment]
I would simply use a single actorId. That actor instance will be unique. As long as your code shares the same constant, it should work.

**Comment by artursouza** [comment]
Additionally, you can also fail the activation of an actor that does not have that "magic" id. This code can be in the Actor class itself.

**Comment by jigargandhi** [comment]
@artursouza  you mentioned discovery problem, however can't we broadcast messages from one daprd and listen from other daprd. This has been implemented for example in Orleans https://dotnet.github.io/orleans/docs/deployment/kubernetes.html

**Comment by artursouza** [comment]
@jigargandhi Yes, it is a possible implementation of this and might remove the need for Placement service. Like I said previously here, the current solution was a middle ground approach to bring redundancy without a completely changing the original approach of relying on Placement service.
